x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryApiKeyIT.java

```diff
@@ -0,0 +1,285 @@
+        assertQuery(API_KEY_ADMIN_AUTH_HEADER,
+            randomBoolean() ? "" : "{\"query\":{\"match_all\":{}}}",
+            apiKeys -> assertThat(apiKeys.size(), equalTo(6)));
+
+        assertQuery(API_KEY_ADMIN_AUTH_HEADER,
+            "{\"query\":{\"bool\":{\"must\":[" +
+                "{\"prefix\":{\"metadata.application\":\"fleet\"}},{\"term\":{\"metadata.environment.os\":\"Cat\"}}]}}}",
+            apiKeys -> {
+                assertThat(apiKeys.size(), equalTo(2));
+                assertThat(apiKeys.get(0).get("name"), oneOf("my-org/ingest-key-1", "my-org/management-key-1"));
+                assertThat(apiKeys.get(1).get("name"), oneOf("my-org/ingest-key-1", "my-org/management-key-1"));
+            }
+        );
+
+        assertQuery(API_KEY_ADMIN_AUTH_HEADER,
+            "{\"query\":{\"terms\":{\"metadata.tags\":[\"prod\",\"east\"]}}}",
+            apiKeys -> {
+                assertThat(apiKeys.size(), equalTo(5));
+            });
+
+        assertQuery(API_KEY_ADMIN_AUTH_HEADER,
+            "{\"query\":{\"range\":{\"creation_time\":{\"lt\":\"now\"}}}}",
+            apiKeys -> {
+                assertThat(apiKeys.size(), equalTo(6));
+            });
+
+        // Search for keys belong to an user
+        assertQuery(API_KEY_ADMIN_AUTH_HEADER,
+            "{ \"query\": { \"term\": {\"username\": \"api_key_user\"} } }",
+            apiKeys -> {
+                assertThat(apiKeys.size(), equalTo(2));
+                assertThat(apiKeys.stream().map(m -> m.get("name")).collect(Collectors.toSet()),
+                    equalTo(Set.of("my-ingest-key-1", "my-alert-key-2")));
+            });
+
+        // Search for keys belong to users from a realm
+        assertQuery(API_KEY_ADMIN_AUTH_HEADER,
+            "{ \"query\": { \"term\": {\"realm_name\": \"default_file\"} } }",
+            apiKeys -> {
+                assertThat(apiKeys.size(), equalTo(6));
+                // search using explicit IDs
+                try {
+
+                    assertQuery(API_KEY_ADMIN_AUTH_HEADER,
+                        "{ \"query\": { \"ids\": { \"values\": ["
+                            + apiKeys.stream().map(m -> "\"" + m.get("id") + "\"").collect(Collectors.joining(",")) + "] } } }",
+                        keys -> {
+                            assertThat(keys.size(), equalTo(6));
+                        });
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
+                }
+            });
+
+        // Search for fields outside of the allowlist fails
+        assertQueryError(API_KEY_ADMIN_AUTH_HEADER, 400,
+            "{ \"query\": { \"prefix\": {\"api_key_hash\": \"{PBKDF2}10000$\"} } }");
+
+        // Search for api keys won't return other entities
+        assertQuery(API_KEY_ADMIN_AUTH_HEADER,
+            "{ \"query\": { \"term\": {\"name\": \"someone\"} } }",
+            apiKeys -> {
+                assertThat(apiKeys, empty());
+            });
+
```

Comment:
Can we also have a test for searching on fields that aren't allowed in the API but are used by the service itself (doc_type, api_key_invalidated)?

Reply:
Thanks for reminding me. I thought about it but got lost in the process. It's now added.